forc-plugins/forc-client/src/util/pkg.rs

```diff
@@ -85,3 +102,366 @@ pub(crate) fn built_pkgs(path: &Path, build_opts: &BuildOpts) -> Result<Vec<Arc<
+        .unwrap_or_else(|| format!("Type{}", type_id))
+}
+
+/// Generates an run external function given the `num_targets`. This function
+/// is later used in the generated loader contract.
+fn generate_run_external(num_targets: usize) -> String {
+    let mut func = String::new();
+
+    func.push_str(&format!("fn run_external{}(", num_targets));
+    for i in 1..=num_targets {
+        if i > 1 {
+            func.push_str(", ");
+        }
+        func.push_str(&format!("load_target{}: ContractId", i));
+    }
+    func.push_str(") -> ! {\n");
+
+    // Generate assembly
+    func.push_str("    asm(\n");
+    for i in 1..=num_targets {
+        func.push_str(&format!("        load_target{}: load_target{},\n", i, i));
+    }
+    for i in 2..=num_targets {
+        func.push_str(&format!("        load_target{}_heap,\n", i));
+    }
+    func.push_str("        heap_alloc_size,\n");
+    for i in 1..=num_targets {
+        func.push_str(&format!("        length{},\n", i));
+    }
+    func.push_str("        ssp_saved,\n");
+    func.push_str("        cur_stack_size,\n");
+    func.push_str("    ) {\n");
+
+    // Get lengths of all chunks
+    for i in 1..=num_targets {
+        func.push_str(&format!("        csiz length{} load_target{};\n", i, i));
+    }
+
+    // Store load_target2 and onwards on the heap
+    for i in 2..=num_targets {
+        func.push_str("        addi heap_alloc_size zero i32;\n");
+        func.push_str("        aloc heap_alloc_size;\n");
+        func.push_str(&format!(
+            "        mcp hp load_target{} heap_alloc_size;\n",
+            i
+        ));
+        func.push_str(&format!("        move load_target{}_heap hp;\n", i));
+    }
+
+    // Save the old $ssp value and shrink the stack
+    func.push_str("        move ssp_saved ssp;\n");
+    func.push_str("        sub cur_stack_size sp ssp;\n");
+    func.push_str("        cfs cur_stack_size;\n");
+
+    // Do the loads
+    func.push_str("        ldc load_target1 zero length1;\n");
+    for i in 2..=num_targets {
+        func.push_str(&format!(
+            "        ldc load_target{}_heap zero length{};\n",
+            i, i
+        ));
+    }
+
+    // Set up jump
+    func.push_str("        addi heap_alloc_size zero i64;\n");
+    func.push_str("        aloc heap_alloc_size;\n");
+    func.push_str("        sw hp ssp_saved i0;\n");
+    func.push_str("    }\n");
+    func.push_str("    __jmp_mem()\n");
+    func.push_str("}\n");
+
+    func
+}
+
+/// Creates a loader contract project, ready to be built at the given path.
+///
+/// Constructs:
+/// - The project folder with a correct structure if it is not already there.
+/// - Loader contract source code, that is created specifically for number of
+///   chunks in hand.
+pub(crate) fn create_chunk_loader_contract(
+    abi: &ProgramABI,
+    chunk_contract_ids: &[String],
+    num_chunks: usize,
+    pkg_name: &str,
+) -> Result<PathBuf> {
+    // Create the proxy contract folder.
```

Comment:
Should we throw an error if `num_chunks` is less than 2? This would also be a good place to check for the upper limit on chunks.